src/System.Text.Json/tests/JsonReaderTests.cs

```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
+        }
 
-        private static List<KeyValuePair<Utf8String, object>> Read(JsonReader json)
+        private static Object ReadObject(ref JsonReader jsonReader)
         {
-            var jsonOutput = new List<KeyValuePair<Utf8String, object>>();
-            var property = new Utf8String("");
-            while (json.Read())
+            var jsonObject = new Object();
+            var jsonMembersList = new List<Members>();
+            List<Pair> jsonPairs = null;
+
+            while (jsonReader.Read())
             {
-                object value;
-                switch (json.TokenType)
+                switch (jsonReader.TokenType)
                 {
+                    case JsonReader.JsonTokenType.Start:
+                        break;
                     case JsonReader.JsonTokenType.ObjectStart:
-                        json.ReadObjectStart();
-                        value = null;
+                        jsonPairs = new List<Pair>();
                         break;
                     case JsonReader.JsonTokenType.ObjectEnd:
-                        json.ReadObjectEnd();
-                        property = new Utf8String("");
-                        value = null;
+                        if (jsonPairs != null)
+                        {
+                            var jsonMembers = new Members {Pairs = jsonPairs};
+                            jsonMembersList.Add(jsonMembers);
+                            jsonObject.Members = jsonMembersList;
+                            return jsonObject;
+                        }
                         break;
                     case JsonReader.JsonTokenType.ArrayStart:
-                        json.ReadArrayStart();
-                        value = null;
                         break;
                     case JsonReader.JsonTokenType.ArrayEnd:
-                        json.ReadArrayEnd();
-                        property = new Utf8String("");
-                        value = null;
                         break;
-                    case JsonReader.JsonTokenType.PropertyName:
-                        property = json.ReadPropertyAsString();
-                        value = null;
+                    case JsonReader.JsonTokenType.Pair:
+                        var pair = new Pair
+                        {
+                            Name = (string) jsonReader.ReadString(),
+                            Value = GetValue(ref jsonReader)
+                        };
+                        if (jsonPairs != null) jsonPairs.Add(pair);
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsString:
-                        value = json.ReadPropertyAsString();
+                    case JsonReader.JsonTokenType.Value:
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsInt:
-                        value = json.ReadPropertyValueAsInt();
+                    case JsonReader.JsonTokenType.Finish:
```

Comment:
You are right. I am already returning false from Read() and we never hit this case anyway.